### ISSUE TITLE
MonaspiceNe Nerd Font Mono ss06 causing characters to overlay on each other

### ISSUE THREAD

**Issue by TehPers** [ryanoasis/nerd-fonts#1503]
#### 🗹 Requirements

- [x] I have searched the [issues](https://github.com/ryanoasis/nerd-fonts/issues) for my issue and found nothing related and/or helpful
- [x] I have searched the [FAQ](https://github.com/ryanoasis/nerd-fonts/wiki/FAQ-and-Troubleshooting) for help
- [x] I have searched the [Wiki](https://github.com/ryanoasis/nerd-fonts/wiki) for help

#### 🎯 Subject of the issue

Experienced behavior:

When enabling `ss06` on MonaspiceNe Nerd Font Mono, characters after consecutive `#`s end up overlayed with the `#`s:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/edcef4f3-3d2c-42d6-b67c-501c412adaa6)

I was able to confirm that MonaspiceNe Nerd Font and MonaspiceNe Nerd Font Propo have the correct behavior.

Screenshot is from vscode, but I was able to reproduce the issue in Windows Terminal as well:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/e8b0b28f-a667-4481-afb2-eedf3fa404d2)

Expected behavior:

The same behavior as Monaspace Neon/MonaspiceNe Nerd Font/MonaspiceNe Nerd Font Propo:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/880ef397-de52-4deb-87e5-477d27b3b595)

For comparison, with `ss06` disabled:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/db9bfd24-230c-44b5-85c6-6277bda2fce5)

Example symbols:

The main symbol I've noticed affected is `#`. I haven't noticed any others, but as far as I know, `ss06` only affects `#`.

```
###$$$$
```

#### 🔧 Your Setup

- _Which font are you using (e.g. `Anonymice Powerline Nerd Font Complete.ttf`)?_
   - `MonaspiceNeNerdFontMono-*.otf`  - as far as I can tell, affects all variants of it. Confirmed with Light and Regular at least
- _Where did you get the file from (download link, self patched, source downloaded from link...)_
   - Releases page (v3.1.1) - `Monaspace.zip`, also tested `Monaspace.tar.xz`
- _Which terminal emulator are you using (e.g. `iterm2`, `urxvt`, `gnome`, `konsole`)?_
   - Windows Terminal, behavior also observed in vscode
- _Are you using OS X, Linux or Windows? And which specific version or distribution?_
   - Windows 11 Pro 22H2

#### ★ Screenshots (Optional)

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/c1cce89d-85f1-4f98-8a6e-6b96d703fba9)

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/969a90ee-1de3-473e-a691-8ee8c12f3584)

`settings.json` for Windows Terminal:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/040ea2a7-951c-44f1-92a4-b51afabdec25)

Copyable settings:

```json
{
    "font": {
        "face": "MonaspiceNe Nerd Font Mono",
        "features": {
            "calt": 1,
            "liga": 1,
            "ss03": 1,
            "ss06": 1,
            "ss08": 1
        }
    }
}
```

**Comment by Finii** [comment]
Thanks for reporting with very good details!
And sorry to hear of your problem.

I believe I instantly know from your description what the problem is:
Unlike all (?) other fonts that have programming ligatures Monaspace uses a different and problematic approach to implement these. I already commented on that fact upstream [1].

The problem I see is that they would need to invest a considerable amount of time to 'fix' the ligatures; it is unclear for me if they will ever get the time for that. The ligs work, somehow, with some limitations, so why spend time and money ...

Let me think a bit about the issue and if there is any generic way we can avoid the problem.
Feel free (you are welcome to) ping me again if I forget to answer within a week or so.

P.S. Do you not see [1a] in VScode? With the non-NerdFontMono patched versions or the unpatched one?

[1a] https://www.github.com/githubnext/monaspace/issues/115
[1b] https://www.github.com/githubnext/monaspace/issues/132

_Links without backlink as this Issue is not really helpful over there_

**Comment by TehPers** [comment]
For 1a, I notice the same behavior that's in that issue when using the non-NFM patched version, so that's definitely an issue with Monaspace. I *do* notice that the behavior is a bit different with the NFM variant though. To fully explain the issue (though this is not a NFM issue, but there are differences between the two variants):

With non-NFM (normal Monaspace Neon), while the cursor appears after the arrow, in reality the cursor is between the `=` and `>` characters, just not visually:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/7c447d6f-8e99-4107-a900-d005994882f2)

When pressing the right arrow key, you actually need to press it twice to pass through the `=>`, one to go past the `=` and the other to go past the `>`, though the cursor *visually* stays at the end of the line here.

With the NFM patched version (MonaspiceNe Nerd Font Mono), while the behavior is almost entirely the same, the difference is that the cursor is now stuck in the middle of the two characters:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/55b1f178-3e11-4f93-8954-32d635fec607)

As for the behavior noted in this issue, I've only noticed that behavior with the NFM variants of the fonts.

If this issue can't be solved from the Nerd Font side, feel free to close the issue. This does seem like an issue with Monaspace itself at the very least.

**Comment by TehPers** [comment]
Adding more context, looks like this happens with other ligatures as well. For example, from `ss03`, the ligature for `<!--` seems to have a similar issue.

Here's from `MonaspiceNe Nerd Font` (`Monaspace Neon` looks the same as this):

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/437a679d-d82f-4843-a92b-946bedf1671e)

Here's from `MonaspiceNe Nerd Font Mono`:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/c6df3c3b-71a0-4b03-aa72-dbef3d0819c0)

**Comment by Finii** [comment]
Well, that all is the problem because Monaspace's idea of how to create the ligs.
I wonder how much work it really is and if I should just change the ligs in Glyph3 and upload a PR at the Monaspace repo.
But then, the repo seems ... not watched by anyone at the moment. Maybe we should try to reach someone and clarify the state.
I guess the Ligs could be fixed within some few days :thinking: 

But then, why should I invest time there, I can not fix all and everything. I already now have no time left.

**Comment by TehPers** [comment]
To be clear - I'm not saying that it needs to be fixed here. Just wanted to report it as an issue, feel free to close as "wont fix" if you feel like it falls into Monaspace's domain to fix.

**Comment by Finii** [comment]
It does indeed.
Well, in some exceptional circumstances we fix bug in sourcefonts locally here, like for example
* #1546 

(which is still not fixed upstream despite the pending PR that they even said they will merge).

I mark this as wontfix as you suggested but keep it open to not forget it. Maybe I have to attend some boring meeting and can work on it then :grimacing:

**Comment by Finii** [comment]
@allcontributors please add @TehPers for bug

**Comment by mihaicris-adoreme** [comment]
I have a similar issue with Iosekva font

installed with macoS brew:
font-iosevka
font-iosevka-nerd-font

Top: Yosevka Fixed
Bottom: Yosevka Nerd Font Mono

<img width="626" alt="image" src="https://github.com/ryanoasis/nerd-fonts/assets/68323356/56a0faea-ae73-4942-973f-a109751ad516">

Where should I address this problem?

Thank you

**Comment by Stealthii** [comment]
@TehPers would you be able to verify if the v1.100 fonts implemented in #1630 resolve all issues you are experiencing?

**Comment by TehPers** [comment]
> @TehPers would you be able to verify if the v1.100 fonts implemented in #1630 resolve all issues you are experiencing?

It's my first time running the font patcher myself, so I'll include the command I ran when checking out the PR branch:

```sh
docker run --rm -v /path/to/nerd-fonts/src/unpatched-fonts/Monaspace:/in:Z -v /path/to/nerd-fonts/patched-fonts/Monaspace:/out:Z nerdfonts/patcher --mono --complete
```

After uninstalling the old font, running the above command, and reinstalling MonaspiceNe Nerd Font Mono (all 8 of the otf files), the `ss06` ligatures seems to be working as expected. Thanks!

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/e3717a32-8a56-43cf-9b4e-792146740997)

**Comment by TehPers** [comment]
@Stealthii I've been using the updated font (MonaspiceNe Nerd Font Mono, generated from your PR), and I found an interesting related issue with a different stylistic set. I know that your PR is only addressing updating Monaspace to v1.100, but just including it here since it seems to be related. This may also be an issue with vscode. To be honest, I'm not entirely sure how to debug where this issue comes from, so I'll include screenshots here, and I can open up an issue on the appropriate repository if anyone has suggestions.

Using `ss07`, if there are three colons next to each other *and* there is red squigglies under what comes immediately after the colons, the cursor steps through the three characters as though the first two are 1.5 wide and the third is 0 wide. This is with the cursor logically between the first and second colon (from the left):

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/9f4f8064-4999-4f2f-a8b6-94337cd78301)

This is without red squigglies:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/43ae6419-6fd6-4245-a131-c27404bd5a18)

It's a bit hard to try to reproduce this with `ss06`'s ligatures, but here's an attempt (which does not exhibit the same problem):

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/cdfa0e7a-c4c4-40dd-8d2a-011457884357)

I'm not really sure why this only happens when there are red squigglies. I couldn't reproduce it with yellow squigglies:

![image](https://github.com/ryanoasis/nerd-fonts/assets/2214247/b7a87e72-c7ac-459b-8a77-b33cd44e97b0)

(If you're trying to reproduce, the first images used Rust, third image was Rust but on the first line as a messed up shebang, and the last one was Markdown)

**Comment by tukykarmakar** [comment]
When using ShureTech Mono font in Konsole, typing anything with 'fl' or 'fi' causes the letter after i or l to overlap on them. For example, the 'l' and 'a' for `flatpak` overlap, and the 'i' and 'g' for `config` overlap. The screenshot below shows some more examples.
![Konsole-fonts-bad-spacing_02](https://github.com/ryanoasis/nerd-fonts/assets/42614682/f38ca48f-5f8f-4fbd-99bb-f09dff5154b6)

I tried using other Nerd Fonts, but it seems like this is happening for ShureTech Mono only. ShureTech Regular and Propo are also having issues with bold letters.

**Comment by Finii** [comment]
> When using ShureTech Mono font in Konsole, typing anything with 'fl' or 'fi' causes the letter after i or l to overlap on them. 

Duplicate of #1631, fixed.

> ShureTech Regular and Propo are also having issues with bold letters.

What does that mean? Probably like #1515? "Bold" letters come in from another font, and there is most likely nothing Nerd Fonts can do. Hmm, well, at least if that is what you mean, report in 1515, so I have all comments in one Issue.

**Comment by MicaelJarniac** [comment]
Possibly related: https://github.com/githubnext/monaspace/issues/190

**Comment by TehPers** [comment]
I should probably mention this, but the original issue I posted seems to be resolved as far as I can tell. MonaspiceNe no longer has the overlapping characters.

I opened an [issue](https://github.com/githubnext/monaspace/issues/216) on Monaspace for the other issue I came across. That does not seem to be a NF issue since it seems to happen with base Monaspace as well.

If we want to continue to use this issue thread for other instances of overlapping characters, then feel free, but I just wanted to make it clear that my issue is resolved at least.
